### PR TITLE
docs: klip-32: A SQL Testing Tool

### PR DESCRIPTION
**Pull request by agavra**
Rendered: https://github.com/agavra/ksql/blob/klip-32/design-proposals/klip-32-sql-testing-tool.md

### CODE REVIEW

File: design-proposals/klip-32-sql-testing-tool.md [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+- _Tombstones_: we will implement `DELETE FROM table WHERE key = value` syntax to allow tombstones
+    to be inserted into tables, both for ksqlDB and the testing tool.
+- _Null Values for Streams_: insert `null` values into streams will not be supported for the first
+    iteration of YATT. Eventually, if we want to support this, we can add a directive like
+    `--@null.value: topic key` to produce a `null` valued record into a specific topic/source.
+- _Other Unsupported Inserts_: there are other types of inserts that ksqlDB doesn't allow for at the
+    moment (e.g. `enum` support or binary formats). This is only somewhat a regression from the 
+    existing QTT tests, so we may or may not decide to support it in YATT.
+    
+### Error Reporting & UX
```
Comment (big-andy-coates) [inline design-proposals/klip-32-sql-testing-tool.md:191]: One thing I'd love to get into QTT, but the large existing set of tests prohibits, is for tests to fail if anything unexpected ends up in the processing logger.  It would be great if we can add this, from the start, to YATT.

Only errors are written to the processing logger.  Test cases should not, generally, be generating such errors. Silently ignoring them means we/users can write tests that pass, yet would spam the processing log (and likely app log) if used in production.  We already know queries that generate a lot of processing errors have terrible performance.

YATT could initially just fail the test case if there are entries in the processing logger. (We'll have to work out how to demux errors when running tests in parallel - probably need to add some kind of UUID or similar somehow).  Later we can add `ASSERT LOG` or similar to assert entries in the processing log, allowing for negative testing. Such negative testing is an important part of QTT.

File: design-proposals/klip-32-sql-testing-tool.md [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+One of the bigger concerns with the existing testing tool is that it's helpful in letting
+us know _that_ an error happened, but not so much _why_ it happened or even at times _what_
+exactly happened. To avoid this pitfall, the errors will include:
+
+- the assert statement that failed
+- the actual data, and the expected data
+- the ability to add `PRINT` statements to debug the issue further by printing topic contents
+
+When data comparisons fail, we will leverage our SQL formatters to display the errors the
+same way we display `SELECT` statements - in a tabular, easy-to-digest way.
```
[the remark this answers is not in the document]
Comment (agavra) [reply design-proposals/klip-32-sql-testing-tool.md:202]: I'll give this some thought and update the KLIP with some ideas - definitely a good point!

File: design-proposals/klip-32-sql-testing-tool.md [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    
+### Error Reporting & UX
+
+One of the bigger concerns with the existing testing tool is that it's helpful in letting
+us know _that_ an error happened, but not so much _why_ it happened or even at times _what_
+exactly happened. To avoid this pitfall, the errors will include:
+
+- the assert statement that failed
+- the actual data, and the expected data
+- the ability to add `PRINT` statements to debug the issue further by printing topic contents
```
Comment (big-andy-coates) [inline design-proposals/klip-32-sql-testing-tool.md:199]: The parse outputs location information when parsing the SQL. So we can also link to the line within the test file that failed, much like QTT does today.  This can save a LOT of time when investigation test failures.

Can we add this to the list please?

File: design-proposals/klip-32-sql-testing-tool.md [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+
+The values will be created in the same way that `INSERT INTO` creates values:
+```sql
+ASSERT DATA source_name [OFFSET offset] ( { column_name } [, ...] ) VALUES ( value [, ...] );
+```
+
+The `ASSERT NO DATA` statement will allow users to ensure that no more data exists in a specified
+source:
+```sql
+ASSERT NO DATA source_name;
```
[the remark this answers is not in the document]
Comment (agavra) [reply design-proposals/klip-32-sql-testing-tool.md:148]: TBH, I hadn't thought about it but definitely a good spec to add; I'll make this explicit :)

File: design-proposals/klip-32-sql-testing-tool.md [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+`ASSERT` is the primary way to ensure conditions in YATT, and will be parsed and implemented 
+using ksqlDB's `AstBuilder` to make sure YATT can leverage the existing expression parsing support. 
+
+At first, `ASSERT` statements will fail if they are sent to a production server, but we may consider
+allowing `ASSERT` in the future in order to allow REPL-based testing/experimentation.
+
+The `ASSERT DATA` statement asserts that data exists at an optionally provided offset (or otherwise 
+sequential based on the last `ASSERT DATA`) with the given values. If the `ASSERT DATA` does not
+specify certain columns they will be ignored (allowing users to assert only subset of the columns
+match what they expect).
```
[the remark this answers is not in the document]
Comment (agavra) [reply design-proposals/klip-32-sql-testing-tool.md:138]: yes, if you look at the example `ROWTIME` is already requested. `OFFSET` is a little special in that it lets you check the ordering of the messages, which is why I bumped it up to a first class citizen

File: design-proposals/klip-32-sql-testing-tool.md [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+    write tests without every leaving the "SQL mindset"
+    
+## Scope
+
+This KLIP covers:
+
+- design of the test file format
+- design of the testing-specific directives
+- outline features that need to be implemented in order to replace QTT/testing-tool
+- sketch out an implementation
```
Comment (big-andy-coates) [inline design-proposals/klip-32-sql-testing-tool.md:57]: I'm not seeing a sketch of how this will be implemented within the KLIP.

Will this be implemented using a fully blown ksqlDB server, running against a (embedded) Kafka, Connect, Schema Registry? Or will it be using the Streams test driver?

If the former, it's likely to be much slower that QTT. Plus, how will the tool support running tests in parallel?  

As discussed offline, ideally, it would support different modes of operation, e.g.
 - test driver vs running against Kafka
 - Using engine directly (like QTT) vs using the rest endpoint (like RQTT)
 - embedded kafka and other services vs  external 

But I'm really not sure this is easily achievable!  So it would be good to know how this initial version will be implemented...

**Comment by MichaelDrogalis** [comment]
Really cool to see this, I think this is exactly where the testing story ought to go. Nice work putting this together!

File: design-proposals/klip-32-sql-testing-tool.md [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+### Extensions
+
+There are some extensions to this testing tool that we may want to consider. I'm listing them here
+because I'm currently in the mindset of thinking about them, and I don't want to forget!
+
+- _BYO Kafka_: we may want to allow users to "plug in their own kafka" cluster and run these tests
+    against a real Kafka cluster. This would allow users to produce whatever data they want
+    outside of this tool, mitigating the limitations described above. It also would allow them
+    to "debug" further when something doesn't go the way they want by examining the input/output
+    topics through ksqlDB.
```
[the remark this answers is not in the document]
Comment (agavra) [reply design-proposals/klip-32-sql-testing-tool.md:213]: `YaaS` (YATT as a Service)?

File: design-proposals/klip-32-sql-testing-tool.md [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+- providing a SQL-driven testing tool will mesh better with our product offering, allowing users to
+    write tests without every leaving the "SQL mindset"
+    
+## Scope
+
+This KLIP covers:
+
+- design of the test file format
+- design of the testing-specific directives
+- outline features that need to be implemented in order to replace QTT/testing-tool
```
[the remark this answers is not in the document]
Comment (agavra) [reply design-proposals/klip-32-sql-testing-tool.md:56]: I'll take a stab at thinking about how we could do this. I imagine it should be somewhat straightforward when we implement BYOK. Then we can just spin up a real ksql server and just pipe things truly e2e

File: design-proposals/klip-32-sql-testing-tool.md [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+
+The intention of this KLIP is to provide a solid foundation for YATT to eventually replace the other
+testing tools and to make sure we are working towards a cleaner code base instead of littering
+it with testing code that needs to be maintained and will eventually slow down our development
+velocity. That being said, it does not cover all the details required, or provide a timeline, to
+migrate the existing tools.
+
+## Design
+
+The best way to describe the testing tool is by motivating example. Below is an example test
```
[the remark this answers is not in the document]
Comment (agavra) [reply design-proposals/klip-32-sql-testing-tool.md:72]: >??

??

File: design-proposals/klip-32-sql-testing-tool.md [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+ASSERT DATA source_name [OFFSET offset] ( { column_name } [, ...] ) VALUES ( value [, ...] );
+```
+
+The `ASSERT NO DATA` statement will allow users to ensure that no more data exists in a specified
+source:
+```sql
+ASSERT NO DATA source_name;
+```
+
+The `ASSERT (TABLE | STREAM)` statement asserts that the given stream or table has the specified
```
[the remark this answers is not in the document]
Comment (agavra) [reply design-proposals/klip-32-sql-testing-tool.md:151]: Is this w.r.t to custom types (e.g. `REGISTER TYPE`)? Or did you mean to `ASSERT STREAM` vs `ASSERT TABLE`? I intended this assert statement to actually require either `STREAM` or `TABLE` (not both) (see example in the motivating example sql file, it asserts a specific source is a stream with schema).

File: design-proposals/klip-32-sql-testing-tool.md (hunk 1 above)
Comment (agavra) [reply design-proposals/klip-32-sql-testing-tool.md:191]: This seems like a good idea, might make migrating tests a little harder so we'd need to be able to disable this unless we wanted to block on adding the correct asserts or fixing up the tests.

Also we could use `ASSERT DATA` on the processing log stream, which should be registered by ksql anyway.

File: design-proposals/klip-32-sql-testing-tool.md (hunk 6 above)
Comment (agavra) [reply design-proposals/klip-32-sql-testing-tool.md:57]: **tl;dr** I'm removing this bullet point from the KLIP 😂 

----

Initially I was going to sketch out an implementation, but I think I might leave that for the PR phase because I'd like to do some more experimentation and I don't want to blow up the scope of (both) this KLIP and the deliverable that I need for query upgrade testing.

My intention is actually to see how bad it is with a full embedded kafka deployment and make a decision based on some "benchmarking". This also involves the least amount of test-specific code, and allows us to reuse as much of the production pipeline as possible (both increasing test coverage, and making things like RQTT migration easy).

I'm somewhat skeptical that the kafka portion of the test was what was slowing it down (I can imagine spinning up a new one for each test would be bad, but if we can reuse an existing cluster I can't imagine that kafka would be a true bottleneck given how fast it can process data).

If I'm wrong, and it proves to be a bottleneck, I'll try to implement a chained topology-test-driver implementation.

File: design-proposals/klip-32-sql-testing-tool.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply design-proposals/klip-32-sql-testing-tool.md]: Just mean I _think_ you're missing the `a`.  super nit.

File: design-proposals/klip-32-sql-testing-tool.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply design-proposals/klip-32-sql-testing-tool.md:162]: The first one: custom types.

File: design-proposals/klip-32-sql-testing-tool.md (hunk 1 above)
Comment (big-andy-coates) [reply design-proposals/klip-32-sql-testing-tool.md:217]: Yeah, something on-by-default.  Can follow the same pattern of other tests in that they should fail if there is unexpected output byond what the test expected.  Unexpected output on the processing log = failed test.

File: design-proposals/klip-32-sql-testing-tool.md (hunk 6 above)
Comment (big-andy-coates) [reply design-proposals/klip-32-sql-testing-tool.md]: My reason for asking for implementation details in the KLIP is because it will strongly implement what's possible and the possible edge cases and issues.

Kafka is a high-throughput, fault-tolerant system. It's not particularly low-latency.  I'd be very surprised if switching to a real Kafka cluster doesn't increase test run time by an order of magnitude. Though feel free to test it.

QTT already takes a fair while to run. It's also key to the way anyone working on the engine works. For example, if I'm working on engine features I'll generally be debugging a single QTT test repeatedly after each code change, running small sets of files less frequently to find the next issue, and the whole suite occasionally. 

It's not just about how long it takes the full suite to run. Though I still think that will be a lot slower. If switching to Kafka means a single test takes longer to run, due to having to start Kafka, then it will _drastically_ reduce developer productivity.  I believe this was the reason the test driver was written in the first place.

IMHO, I would restrict this KLIP to implementation using the test driver only. This would allow us to replace QTT and the ksqlDB testing tool once YATT is feature complete enough.  BYOK and RQTT can be mentioned as future work only, requiring additional KLIPs.

Comment (agavra) [reply design-proposals/klip-32-sql-testing-tool.md]: I tested this and you are right :D i'll restrict the implementation to test-driver

Comment (agavra) [reply design-proposals/klip-32-sql-testing-tool.md]: Updated in the KLIP under `implementation`

File: design-proposals/klip-32-sql-testing-tool.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (agavra) [reply design-proposals/klip-32-sql-testing-tool.md:162]: Added

File: design-proposals/klip-32-sql-testing-tool.md [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+#### `ASSERT DATA`
+
+The `ASSERT DATA` statement asserts that data exists at an optionally provided offset (or otherwise 
+sequential based on the last `ASSERT DATA`) with the given values. If the `ASSERT DATA` does not
+specify certain columns they will be ignored (allowing users to assert only subset of the columns
+match what they expect).
+
+The values will be created in the same way that `INSERT INTO` creates values:
+```sql
+ASSERT DATA source_name [OFFSET at_offset] ( { column_name } [, ...] ) VALUES ( value [, ...] );
```
Comment (vcrfxia) [inline design-proposals/klip-32-sql-testing-tool.md:145]: What about asserting data in internal topics (e.g., repartition topics) as some QTT tests currently do?

File: design-proposals/klip-32-sql-testing-tool.md [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+There are some features that we need to implement in ksqlDB's SQL language in order to put YATT
+on par with the existing testing tools. Some of those are outlined here:
+
+- _Windowed Keys_: when we have structured key support, we can `ASSERT DATA` and just provide the
+    struct as the value. Until then, windowed keys will either be unsupported or stringified.
+- _Tombstones_: we will implement `DELETE FROM table WHERE key = value` syntax to allow tombstones
+    to be inserted into tables, both for ksqlDB and YATT.
+- _Null Values for Streams_: insert `null` values into streams will not be supported for the first
+    iteration of YATT. Eventually, if we want to support this, we can add a directive like
+    `--@null.value: topic key` to produce a `null` valued record into a specific topic/source.
```
Comment (vcrfxia) [inline design-proposals/klip-32-sql-testing-tool.md:216]: Is this not something we want to eventually support outside of testing purposes? Either (1) there are no valid use cases for inserting nulls into streams, in which case why are we testing it, or (2) there are valid use cases for inserting nulls into streams, and we should support it in ksqlDB.

File: design-proposals/klip-32-sql-testing-tool.md [hunk 13]
```diff
@@ -0,0 +1,284 @@
...
+    
+### Error Reporting & UX
+
+One of the bigger concerns with the existing ksql-test-runner is that it's helpful in letting
+us know _that_ an error happened, but not so much _why_ it happened or even at times _what_
+exactly happened. To avoid this pitfall, the errors will include:
+
+- the assert statement that failed, with a link to the line that failed in the test file
+- the actual data, and the expected data
+- the ability to add `PRINT` statements to debug the issue further by printing topic contents
```
Comment (vcrfxia) [inline design-proposals/klip-32-sql-testing-tool.md:229]: Does this mean repurposing the existing print topic syntax for use in the testing tool? This functionality makes a lot of sense but I wonder if repurposing existing syntax for something else will become confusing or problematic if we later add support for push queries/pull queries/print topic statements in the testing tool.

File: design-proposals/klip-32-sql-testing-tool.md [hunk 14]
```diff
@@ -0,0 +1,284 @@
...
+
+-- the below operation is invalid because "col2" will be missing from the
+-- schema after renaming it to col3
+CREATE STREAM bar AS SELECT id, col1, col2 FROM foo;
+CREATE OR REPLACE STREAM bar AS SELECT id, col1, col2 AS col3 FROM foo;
+```
+
+### Test Structure
+
+YATT will accept as a parameter a directory containing testing files, and will run all the
```
Comment (vcrfxia) [inline design-proposals/klip-32-sql-testing-tool.md:113]: Should we also support passing in a single test file directly, so users can debug individual tests/test files without having to run an entire directory of tests?

File: design-proposals/klip-32-sql-testing-tool.md (hunk 11 above)
Comment (agavra) [reply design-proposals/klip-32-sql-testing-tool.md:145]: I discussed this with @big-andy-coates and we came up with two conclusions:

1. this shouldn't be necessary anymore because the historical tests capture all internal topic data and assert that it's the same, so you shouldn't need to manually assert that it stays the same
1. if you _really_ wanted to, you could declare it as a stream/table and use the `ASSERT DATA` calls

File: design-proposals/klip-32-sql-testing-tool.md (hunk 12 above)
Comment (agavra) [reply design-proposals/klip-32-sql-testing-tool.md:216]: That's a good question, but even if we decide there are no valid use cases for insert nulls into streams it's still valid to test what happens if your source (bad) data has nulls in your streams.

File: design-proposals/klip-32-sql-testing-tool.md (hunk 13 above)
Comment (agavra) [reply design-proposals/klip-32-sql-testing-tool.md:229]: I'm not entirely sure I understand the concern?

File: design-proposals/klip-32-sql-testing-tool.md (hunk 14 above)
Comment (agavra) [reply design-proposals/klip-32-sql-testing-tool.md:113]: yup! I will add this into the KLIP

File: design-proposals/klip-32-sql-testing-tool.md (hunk 12 above)
Comment (vcrfxia) [reply design-proposals/klip-32-sql-testing-tool.md:216]: That's fair but it seems weird to add special support for this particular case of "bad data" in the testing tool, since there are other types of bad data we can't produce using SQL syntax either (garbage bytes, malformed data, etc). The KLIP already contains a proposal for BYO Kafka to allow for testing these scenarios. Why is the case of null values in streams special / different?

Comment (agavra) [reply design-proposals/klip-32-sql-testing-tool.md:216]: the big one is parity with QTT/ksql-test-runner; they can't produce garbage bytes or malformed data either. BYOK is a pretty heavy extension if we wanted to bring this up to parity

File: design-proposals/klip-32-sql-testing-tool.md (hunk 13 above)
Comment (vcrfxia) [reply design-proposals/klip-32-sql-testing-tool.md:229]: My original thinking was: If YATT is eventually extended to include support for the current RQTT, then YATT will have to support statements for push and pull queries. If it's eventually extended to also support `PRINT <topic>;` statements (a natural extension to push query support), but the `PRINT <topic>;` syntax already has a specific meaning for the testing tool (i.e., print contents of the topic from the beginning) which conflicts with the meaning of the KSQL statement (print newly arriving data only), then it'll be confusing what the expected behavior is. To remedy the confusion we could:
- have YATT assume `auto.offset.reset=earliest` by default, so the KSQL statement `PRINT <topic>;` also prints from the beginning
- require KSQL syntax for printing topics in YATT, e.g., `PRINT 'foo';` complete with single quotes and everything. (I assume this was already the plan, in order to reuse the parser.)

What are your thoughts on the former (defaulting to `auto.offset.reset=earliest`)? While reading the proposal I had in my mind we wouldn't do that, though the KLIP wasn't explicit one way or the other.

Comment (agavra) [reply design-proposals/klip-32-sql-testing-tool.md:229]: ah, my intention was actually to have YATT's `PRINT` statement match the semantics of KSQL's `PRINT` statement - the only difference would be that it knows when the topic ends and it doesn't hang forever (to be automation friendly) and instead ends on topic end. 

That means that if you wanted to print from beginning you would need `PRINT 'foo' FROM BEGINNING;`.

Comment (vcrfxia) [reply design-proposals/klip-32-sql-testing-tool.md:229]: Got'cha, that makes sense to me then. Thanks!

**Comment by PeterLindner** [comment]
@agavra will ```RUN SCRIPT``` be supported too? I think it would be very useful to test the actual scripts that will be deployed via CI/CD from within the test files, eg:
```sql
-- createFoo.ksql
CREATE STREAM IF NOT EXISTS foo WITH (...);
```
```sql
-- createBar.ksql
CREATE OR REPLACE STREAM BAR 
WITH (...) 
AS SELECT * FROM FOO;
```
```sql
-- test.ksql
--@test: foobar
RUN SCRIPT 'createFoo.ksql';
ASSERT STREAM FOO <props>;
INSERT INTO foo <stuff>;

RUN SCRIPT 'createBar.ksql';
ASSERT STREAM bar <props>;
ASSERT DATA bar <stuff>; 
```

**Comment by agavra** [comment]
@PeterLindner  - yes! That was exactly the intention, and I was just in the process of updating the KLIP to describe that in the context of CI/CD pipelines. GMTA
